### PR TITLE
logging: cAVS HDA based logger

### PR DESCRIPTION
**Pull request by teburd**
Adds a log backend that maintains a ringbuffer in coordination
with cAVS HDA.

The DMA channel is setup by the application and this acts as a in memory buffer logger (without overwrite) until that is done.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

### CODE REVIEW

File: soc/xtensa/intel_adsp/tools/cavstool.py [hunk 1]
```diff
@@ -551,6 +555,19 @@ def ipc_command(data, ext_data):
         for i in range(0, 256):
             buf[i] = i
         hda_streams[stream_id].write(buf)
+    elif data == 12: # HDA PRINT
+        log.info("Doing HDA Print")
+        stream_id = ext_data & 0xFF
+        buf_len = ext_data >> 8 & 0xFFFF
+        hda_str = hda_streams[stream_id]
+        #hda_str.mem.seek(0)
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/tools/cavstool.py:563]: Leftover this commented line ?

File: soc/xtensa/intel_adsp/tools/cavstool.py [hunk 2]
```diff
@@ -551,6 +555,19 @@ def ipc_command(data, ext_data):
...
+        pos = hda_str.mem.tell()
+        buf_data = hda_str.mem.read(buf_len).decode("utf-8", "replace")
+        log.info(f"DSP LOG MSG (idx: {pos}, len: {buf_len}): {buf_data}")
+        pos = hda_str.mem.tell()
+        if pos >= hda_str.buf_len*2:
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/tools/cavstool.py:568]: What is the logic here, why the check is for greater than twice buffer's length ?

File: subsys/logging/log_backend_cavs_hda.c [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+	if (ret != 0) {
+		ret = 0;
+		goto exit;
+	}
+
+	/* The hardware tells us what space we have available, and where to
+	 * start writing. If the buffer is full we have no space.
+	 */
+	if (status.free <= 0) {
+		ret = 0;
```
Comment (ceolin) [inline subsys/logging/log_backend_cavs_hda.c:75]: This callback is really weird, the doc says that it should return the number of bytes processed, dropped or discarded. Don't know how this should be done, if write 10 bytes out of 15 what should you return ? 10 or -5 ? Anyway, I think return 0 is fine, -ENOSPC also would be ok.

File: subsys/logging/log_backend_cavs_hda.c [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+		goto exit;
+	}
+
+	/* Account for buffered writes since last dma_reload */
+	uint32_t available =  status.free - hda_log_buffered;
```
Comment (ceolin) [inline subsys/logging/log_backend_cavs_hda.c:80]: Can't you have an underflow here ?

**Review by ceolin — DISMISSED** [review]

File: soc/xtensa/intel_adsp/tools/cavstool.py (hunk 2 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/tools/cavstool.py:568]: HDAStream (class) actually has 2 buffers of buf len right next to each other, these get flip/flopped as it reads things in. So this maybe is wrong, but it lets the strings be read in

File: subsys/logging/log_backend_cavs_hda.c (hunk 3 above)
Comment (teburd) [reply subsys/logging/log_backend_cavs_hda.c:75]: yeah I wasn't sure either, 0 seems to work out ok, I'll look closer at it. 0 is probably wrong though

File: subsys/logging/log_backend_cavs_hda.c (hunk 4 above)
Comment (teburd) [reply subsys/logging/log_backend_cavs_hda.c:80]: I don't believe so, I'll look closer though

File: soc/xtensa/intel_adsp/tools/cavstool.py (hunk 1 above)
Comment (teburd) [reply soc/xtensa/intel_adsp/tools/cavstool.py:563]: Yes, thanks

File: subsys/logging/log_backend_cavs_hda.c (hunk 3 above)
Comment (teburd) [reply subsys/logging/log_backend_cavs_hda.c]: Looking at how its used, I always need to return all the bytes here I believe since I'm not really excited to try and write some partial number of bytes. So I always return the full length of the message now.

File: subsys/logging/log_backend_cavs_hda.c (hunk 4 above)
Comment (teburd) [reply subsys/logging/log_backend_cavs_hda.c]: I'm confident this will never underflow, however an __ASSERT is added as cheap insurance.

File: subsys/logging/log_backend_cavs_hda.c [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+		.source_data_size = 4,
+	};
+
+	res = dma_config(hda_log_dev, hda_log_chan, &hda_log_dma_cfg);
+	__ASSERT(res == 0, "DMA config failed");
+
+	res = dma_start(hda_log_dev, hda_log_chan);
+	__ASSERT(res == 0, "DMA start failed");
+
+	k_timer_init(&hda_log_timer, hda_log_periodic, NULL);
```
Comment (nashif) [inline subsys/logging/log_backend_cavs_hda.c:225]: do we need this here, shouldn't this be handled by the core logging infrastructure? @dcpleung  can you please confirm?

Comment (dcpleung) [reply subsys/logging/log_backend_cavs_hda.c:225]: The logging subsys doesn't have an explicit way of forcing a transfer. The `log_output_flush()` simply calls the backend's output function to send the partially filled buffer to the output. There is no "end of message' notification to the backend. If there is a large enough buffer, you can wait till the buffer is filled with one log message, and start the DMA transfer when the backend output function is called. However, in logging deferred mode, multiple log messages may be processed at the same time, and you will have a series of DMA transfer requests.

Comment (teburd) [reply subsys/logging/log_backend_cavs_hda.c:225]: Right so the way this is done here with a watermark and timer to ensure a periodic flush (when the watermark doesn't get hit) makes sense then? I can imagine maybe in panic mode I should always flush perhaps then, with some 0 byte padding maybe? I was afraid of padding the output because I wasn't sure if 0 byte padding sys-t dictionary or sys-t formatting would screw things up, but that would be needed to get immediate dma writes all the time as there's a minimum 128 byte transfer

Comment (dcpleung) [reply subsys/logging/log_backend_cavs_hda.c:225]: Padding would probably not work with dictionary and sys-t catalog messages. I think you can do a DMA transfer on a partially filled block. Let the next log message fill the rest of the block, and do another transfer of the same block. Though it would require the log streamer on the other side to be aware of this.

File: subsys/logging/Kconfig.backends [hunk 6]
```diff
@@ -325,6 +325,48 @@ source "subsys/logging/Kconfig.template.log_format_config"
...
+	int "Size of ring buffer"
+	range 128 8192
+	default 4096
+	help
+	  Sets the ring buffer size cAVS HDA uses for logging. Effectively
+	  determines how many log messages may be written to in a period of time.
+	  The period of time is decided by how often to inform the hardware of
+	  writes to the buffer.
+
+config LOG_BACKEND_CAVS_HDA_CHANNEL
```
Comment (abonislawski) [inline subsys/logging/Kconfig.backends:352]: Can we setup hda channel in runtime?
In SOF host will send IPC with channel info

Comment (kv2019i) [reply subsys/logging/Kconfig.backends:352]: In SOF, we currently allow lazy alloc of the HDA DMA backend. Tracing can be started, but as the DMA channel is not yet known, nothing is sent out. This is nice to gather very early logs in the buffers. At one point in boot, the host side kernel sends IPC that configures DMA trace and passes the channel. This comes to SOF in: https://github.com/thesofproject/sof/blob/main/src/ipc/ipc3/handler.c#L796
And ends up configuring the DMA here:
https://github.com/thesofproject/sof/blob/main/src/trace/dma-trace.c#L463

Something similar would be nice for this backend as well.

Reserving a fixed channel is problematic as 1) at least in theory. we could have usages where we want all DMA channels for non-log use, 2) we have large number of old Linux kernel versions deployed that expect SOF FW ABI to allow to dynamically configure the DMA trace/log channel. Requiring a fix channel would break all these old kernels and getting a kernel change deployed (to all distros and users) is much slower and complicated than distributing a new FW. So in short, ABI compatibility is sacred in SOF.

File: subsys/logging/log_backend_cavs_hda.c [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+	if (!atomic_cas(&hda_log_lock, 0, 1)) {
+		return;
+	}
+
+	uint32_t nearest128 = hda_log_buffered & ~((128) - 1);
+
+	if (nearest128 > 0) {
+		hda_log_buffered = hda_log_buffered - nearest128;
+
+		dma_reload(hda_log_dev, hda_log_chan, 0, 0, nearest128);
```
Comment (kv2019i) [inline subsys/logging/log_backend_cavs_hda.c:46]: I was about to make a comment about avoiding a magic number with 128, but upon second thought, maybe this is more readable after all.

**Review by kv2019i — COMMENTED** [review]
Thanks @teburd ! This looks pretty nice. A few questions inline. This is already useful now, so I'm ok to do extensions to the "IPC hook" interface in separate PRs if this otherwise looks good.

File: subsys/logging/log_backend_cavs_hda.c (hunk 7 above)
Comment (kv2019i) [reply subsys/logging/log_backend_cavs_hda.c:46]: And for context, in Linux SOF driver, we acquire and set up HDA DMA trace here:

Generic code that sets up kernel side and sends the IPC:
https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc3-dtrace.c#L379

Above calls to snd_sof_dma_trace_init() which handles the HW/DSP specific actions to set up the DMA hw.
 https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda-trace.c#L39

File: subsys/logging/Kconfig.backends (hunk 6 above)
Comment (teburd) [reply subsys/logging/Kconfig.backends]: Fixed, the logger is now autostarted and fills the buffer with log messages with a "keeps oldest" as the behavior, it now requires a channel param be passed in

File: soc/xtensa/intel_adsp/tools/cavstool.py [hunk 8]
```diff
@@ -551,6 +555,18 @@ def ipc_command(data, ext_data):
         for i in range(0, 256):
             buf[i] = i
         hda_streams[stream_id].write(buf)
+    elif data == 12: # HDA PRINT
```
Comment (andyross) [inline soc/xtensa/intel_adsp/tools/cavstool.py:558]: Would prefer to see the protocol handling, which is a test thing very unlikely to be used by production code (SOF has its own much more elaborate IPC protocol) cleanly separated from the driver here in a different patch.

File: subsys/logging/Kconfig.backends [hunk 9]
```diff
@@ -325,6 +325,41 @@ source "subsys/logging/Kconfig.template.log_format_config"
...
+	  writes to the buffer.
+
+config LOG_BACKEND_CAVS_HDA_FLUSH_TIME
+	int "Time in milliseconds to periodically flush writes to hardware"
+	range 1 10000
+	default 500
+	help
+	    The cAVS HDA backend periodically writes out its buffer contents
+	    to hardware by informing the DMA hardware the contents of the ring
+	    buffer.
```
Comment (andyross) [inline subsys/logging/Kconfig.backends:359]: Not sure a timer is really the right thing here, that's overhead inside the device when really decisions about output flushing want to be part of the app.    If nothing else, you could make it configurable by e.g. treating "0" as meaning "disabled".  I think the watermark-based capacity flushing should be acceptable for basically all routine use, and if we "Need To Get Stuff Out Now" we're probably in some kind of error condition and want a facility to do a synchronous flush anyway.

File: subsys/logging/log_backend_cavs_hda.c [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+	atomic_set(&hda_log_lock, 0);
+}
+
+
+static int hda_log_out(uint8_t *data, size_t length, void *ctx)
+{
+	int ret;
+	struct dma_status status;
+
+	while (!atomic_cas(&hda_log_lock, 0, 1)) {
```
Comment (andyross) [inline subsys/logging/log_backend_cavs_hda.c:63]: I was on board with this thing above when you had an opportunistic lock where you could fail.  But here you're spinning anyway.  Let's not reinvent things, use a proper spinlock.  If we really have latency concerns that want the opportunistic algorithm above, then let's add that as as "try_lock" primitive to spinlocks instead.

File: subsys/logging/log_backend_cavs_hda.c [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+	}
+
+	/* Defaults when DMA not yet initialized */
+	uint32_t dma_free = sizeof(hda_log_buf);
+	uint32_t write_pos = 0;
+
+	if (initialized) {
```
Comment (andyross) [inline subsys/logging/log_backend_cavs_hda.c:70]: Why test this?  It's not unreasonable for a driver API to assume its own initialization code ran.  If this it to handle some kind of "log happens too early" case, then presumably you need to adjust the device init priority?

File: subsys/logging/log_backend_cavs_hda.c [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+		ret = dma_get_status(hda_log_dev, hda_log_chan, &status);
+		if (ret != 0) {
+			ret = length;
+			goto exit;
```
Comment (andyross) [inline subsys/logging/log_backend_cavs_hda.c:74]: Don't get this bit.  This is treating an error in dma_get_status (in a device that wasn't even initialized!) as... success and claiming to have accepted all the data correctly?

File: subsys/logging/log_backend_cavs_hda.c (hunk 5 above)
Comment (andyross) [reply subsys/logging/log_backend_cavs_hda.c:257]: Just to pile on from a comment elsewhere (that I never pushed, sorry).  I don't think the timer is doing much good here.  Let the app decide on flushing policy (creating a period timer to call flush() in the app is really trivial).  And if we do want a timer here, it should probably leverage its better access for more intelligence: have reset itself when we write to a an empty-and-idle buffer, such that the budget becomes the "most time a byte of output will sit unflushed" and we promise no spurious/tiny output except when really needed.

File: tests/boards/intel_adsp/hda_log/src/logger.c [hunk 13]
```diff
@@ -0,0 +1,67 @@
+/* Copyright (c) 2022 Intel Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Comment (andyross) [inline tests/boards/intel_adsp/hda_log/src/logger.c:3]: Similar content to elsewhere: put the test code in a separate patch from the driver.  Especially with this feature, the protocol needed to do the management is wildly different from what SOF will use.

File: tests/boards/intel_adsp/hda_log/src/logger.c [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+	/*  Now send the next one */
+	do {
+		done = cavs_ipc_send_message(CAVS_HOST_DEV, IPCCMD_HDA_PRINT,
+					     (written << 8) | CHANNEL);
+	} while (!done);
+}
+
+
+void test_hda_logger(void)
+{
```
Comment (andyross) [inline tests/boards/intel_adsp/hda_log/src/logger.c:43]: Test coverage pedantry: Not seeing a test case that exercises watermark flush or the overflow conditions.

File: subsys/logging/log_backend_cavs_hda.c [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+		memcpy(&hda_log_buf[0], &data[copy_len], wrap_copy_len);
+	}
+
+	ret = length;
+	hda_log_buffered += length;
+
+exit:
+	atomic_set(&hda_log_lock, 0);
+
+	if ((hda_log_buffered > sizeof(hda_log_buf)/2 || panic_mode) && initialized) {
```
Comment (andyross) [inline subsys/logging/log_backend_cavs_hda.c:133]: This is a race: you're testing outside the lock, so two simultaneous threads trying to test it might flush twice, or plausibly miss it entirely if "panic_mode" or "initialized" get mutated.  Best practice is always to put all access to shared data inside a single lock and be consistent about it.   In practice it looks like the flush is just going to take the same lock anyway, so errors would be weird and very rare if they happened.

Also: you want a kconfig to tune that watermark.  Logging is bursty.  It's possible there are some app configurations that need to start flushing almost immediately so that they have more than half the buffer avaialable for a big stack dump, etc....  This is the kind of number that never works for everyone, someone's going to need to hack it at some point.

File: subsys/logging/log_backend_cavs_hda.c [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+
+	res = dma_start(hda_log_dev, hda_log_chan);
+	__ASSERT(res == 0, "DMA start failed");
+
+	k_timer_init(&hda_log_timer, hda_log_periodic, NULL);
+	k_timer_start(&hda_log_timer,
+		      K_MSEC(CONFIG_LOG_BACKEND_CAVS_HDA_FLUSH_TIME),
+		      K_MSEC(CONFIG_LOG_BACKEND_CAVS_HDA_FLUSH_TIME));
+
+	initialized = true;
```
Comment (andyross) [inline subsys/logging/log_backend_cavs_hda.c:262]: Pedantic: the k_timer_start() call is scheduling a function to run on an uninitialized object here.  You want to flag it initialized before you call API functions on it, or even schedule API functions.  (Obviously in practice this won't matter)

**Review by andyross — COMMENTED** [review]
Sorry, I realized I had never pushed out a bunch of these comments.  Also added a few new ones.

On the whole it looks great.  Some general nitpicks which would be nice to see fixed.  In particular the home-brew spinlock really needs to be a real one, and the "initialized" flag seems like a code smell thing that would be better treated by control over initialization order (log backends do indeed need to be initialized very early).

**Comment by teburd** [comment]
> Sorry, I realized I had never pushed out a bunch of these comments. Also added a few new ones.
> 
> On the whole it looks great. Some general nitpicks which would be nice to see fixed. In particular the home-brew spinlock really needs to be a real one, and the "initialized" flag seems like a code smell thing that would be better treated by control over initialization order (log backends do indeed need to be initialized very early).

I have no control over the initialization of this, as in SoF its actually initialized by an IPC request from linux module.

spinlocks have no try_lock method at the moment, which would resolve the issue. I don't actually want to block when attempting to flush.

File: subsys/logging/log_backend_cavs_hda.c (hunk 11 above)
Comment (teburd) [reply subsys/logging/log_backend_cavs_hda.c:70]: This is because the DMA channel may not be setup yet, that happens when SoF decides to call the init function stemming from an IPC message from the kernel, which happens... probably when sof-logger is started

File: subsys/logging/log_backend_cavs_hda.c (hunk 5 above)
Comment (teburd) [reply subsys/logging/log_backend_cavs_hda.c:257]: This is coming directly from the way dma-trace works in sof, https://github.com/thesofproject/sof/blob/main/src/trace/dma-trace.c#L492

**Comment by andyross** [comment]
> spinlocks have no try_lock method at the moment, which would resolve the issue. I don't actually want to block when attempting to flush.

You don't need a try_lock though, and spinlocks don't ever block (because they hold the cpu).  Take the lock, see that the DMA is in progress, release the lock.  I really don't think DIY synchronization is the answer here.

File: subsys/logging/log_backend_cavs_hda.c (hunk 16 above)
Comment (teburd) [reply subsys/logging/log_backend_cavs_hda.c:262]: This is a fair point, initialized is meant to signify the DMA channel is available to be flushed to. dma_ready might be a better variable name for this, and moving it before the timer makes sense

**Comment by andyross** [comment]
> I have no control over the initialization of this, as in SoF its actually initialized by an IPC request from linux module.

This is a Zephyr log backend, though.  It needs to act like a log backend and thus be available to the logging API very early.  Maybe what you really want is a working/"initialized" log backend that exists before it gets assigned an HDA channel?  That is, the channel and DMA state becomes runtime state and not a property of "initialization"?

File: subsys/logging/log_backend_cavs_hda.c (hunk 11 above)
Comment (andyross) [reply subsys/logging/log_backend_cavs_hda.c:70]: OK, I think I'm understanding.  This logic actually looks OK, what threw me is the use of the word "initialized" here.  The log backend actually can be "initialized" as a memory buffer and accept output before getting a hardware backend set up.  How about testing your channel state instead?  That is, instead of "if(initialized)" you write "if(dma_available())" or whatever?

File: subsys/logging/log_backend_cavs_hda.c (hunk 12 above)
Comment (andyross) [reply subsys/logging/log_backend_cavs_hda.c:74]: OK.  But now I have another question: why is it valid to call dma_get_status() here if initialized is unset (which we now know means "the device hasn't been assigned an HDA channel by the host driver")?  Still not understanding.

**Comment by teburd** [comment]
> > I have no control over the initialization of this, as in SoF its actually initialized by an IPC request from linux module.
> 
> This is a Zephyr log backend, though. It needs to act like a log backend and thus be available to the logging API very early. Maybe what you really want is a working/"initialized" log backend that exists before it gets assigned an HDA channel? That is, the channel and DMA state becomes runtime state and not a property of "initialization"?

It is available but won't write with DMA until initialized, which is the whole intent of that variable.

**Comment by andyross** [comment]
> It is a working and available log backend, but won't write with DMA until initialized, which is the whole intent of that variable.

I understand that now.  I'm saying that I was confused because in traditional usage in Zephyr, "initialized" means that a driver has finished its initialization function and that its API can be used.  It is not a statement of runtime state, which is what it means here.  I strongly recommend renaming to something that addresses the ambiguity, but that's not worth a -1.

**Comment by teburd** [comment]
> > It is a working and available log backend, but won't write with DMA until initialized, which is the whole intent of that variable.
> 
> I understand that now. I'm saying that I was confused because in traditional usage in Zephyr, "initialized" means that a driver has finished its initialization function and that its API can be used. It is not a statement of runtime state, which is what it means here. I strongly recommend renaming to something that addresses the ambiguity, but that's not worth a -1.

It's a poorly named variable in this context, will fix :+1:

**Comment by teburd** [comment]
I've added the spin locks but it now seems to dead lock, the timer callback tries to take the lock and spins forever now from what I can tell

**Comment by teburd** [comment]
I've added the spin locks but it now seems to dead lock, the timer callback tries to take the lock and spins forever now from what I can tell

File: subsys/logging/log_backend_cavs_hda.c (hunk 12 above)
Comment (teburd) [reply subsys/logging/log_backend_cavs_hda.c]: this is why dma_get_status was wrapped in if (initialized)

**Comment by teburd** [comment]
I had to move the hook outside of the spin lock to avoid the deadlock

**Comment by dcpleung** [comment]
This might help with "dma not ready": https://github.com/zephyrproject-rtos/zephyr/pull/43405

File: subsys/logging/Kconfig.backends [hunk 17]
```diff
@@ -325,6 +325,41 @@ source "subsys/logging/Kconfig.template.log_format_config"
 
 endif # LOG_BACKEND_ADSP
 
+config LOG_BACKEND_CAVS_HDA
+	bool "cAVS HDA backend"
+	depends on SOC_FAMILY_INTEL_ADSP && DMA && DMA_CAVS_HDA
```
Comment (dcpleung) [inline subsys/logging/Kconfig.backends:330]: Since you only implement the `.process` API, this needs to depend on `CONFIG_LOG2` also.

Comment (teburd) [reply subsys/logging/Kconfig.backends:330]: When can we delete v1?

Comment (dcpleung) [reply subsys/logging/Kconfig.backends]: Probably after another release? Deprecated stuff needs to stay for 2 releases before they can be deleted.

**Review by ceolin — DISMISSED** [review]

File: subsys/logging/log_backend_cavs_hda.c (hunk 10 above)
Comment (teburd) [reply subsys/logging/log_backend_cavs_hda.c]: Done! did cause me a minor headache with the hook, but sorted that out

**Comment by teburd** [comment]
Added some additional commentary to clarify a few things

File: subsys/logging/log_backend_cavs_hda.c [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+	 * the nearest 128 divisible value of hda_log_buffered,
+	 * where hda_log_buffered is then subtracted by the same amount.
+	 * After which status.free should only increase in value.
+	 *
+	 * Assert this trueth though, just in case.
+	 */
+	__ASSERT_NO_MSG(dma_free > hda_log_buffered);
+	uint32_t available = dma_free - hda_log_buffered;
+
+	/* If there isn't enough space for the message there's an overflow */
```
Comment (andyross) [inline subsys/logging/log_backend_cavs_hda.c:92]: It occurs to me that this is a situation where you might plausibly want a short write kind of behavior, returning the partial number of bytes consumed.  I have no idea if the log layer is prepared to handle that sort of thing cleanly, but I know it will emit "N messages dropped" kind of diagnostics on the other side (insertion into the async queue), so it's a feature the subsystem might evolve toward in the future.

File: tests/boards/intel_adsp/hda_log/testcase.yaml [hunk 19]
```diff
@@ -0,0 +1,7 @@
+tests:
+  boards.intel_adsp:
+    platform_allow: intel_adsp_cavs15 intel_adsp_cavs18 intel_adsp_cavs20 intel_adsp_cavs25
+  boards.intel_adsp.1cpu:
+    platform_allow: intel_adsp_cavs15 intel_adsp_cavs18 intel_adsp_cavs20 intel_adsp_cavs25
+    extra_configs:
+      - CONFIG_MP_NUM_CPUS=1
```
Comment (andyross) [inline tests/boards/intel_adsp/hda_log/testcase.yaml:7]: Why 1 cpu?  That sounds suspicious, since obviously this needs to work in SMP environments.

**Review by andyross — DISMISSED** [review]
This looks clean to me.  A few more notes that I noticed, and I still wouldn't cry to see the test code split out of the main patch (especially since it uses a protocol that has nothing to do with SOF), but those aren't worth -1's.

File: tests/boards/intel_adsp/hda_log/testcase.yaml (hunk 19 above)
Comment (teburd) [reply tests/boards/intel_adsp/hda_log/testcase.yaml:7]: The real goal was to test with and without KERNEL_COHERENCY just like the hda tests do

Comment (teburd) [reply tests/boards/intel_adsp/hda_log/testcase.yaml:7]: Which of course I forgot to do the flush on the buffer when its not enabled

File: subsys/logging/log_backend_cavs_hda.c [hunk 20]
```diff
@@ -0,0 +1,292 @@
...
+static atomic_t hda_log_flags;
+
+static uint32_t hda_log_flush(void)
+{
+	uint32_t nearest128 = hda_log_buffered & ~((128) - 1);
+
+	if (nearest128 > 0) {
+		hda_log_buffered = hda_log_buffered - nearest128;
+#if !(IS_ENABLED(CONFIG_KERNEL_COHERENCE))
+		z_xtensa_cache_flush(hda_log_buf, CONFIG_LOG_BACKEND_CAVS_HDA_SIZE);
```
Comment (andyross) [inline subsys/logging/log_backend_cavs_hda.c:42]: This came up via offline discussion, but FWIW: it's probably a better idea to make the hda_log_buf uncached always (by running the pointer through z_xtensa_uncached_ptr(), which works at runtime (though of course in the case where the pointer is already uncached will compile out as a noop).  That avoids the need to have two test cases.

**Review by andyross — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Review by kv2019i — APPROVED** [review]

**Review by laurenmurphyx64 — APPROVED** [review]

**Review by yerabolu — APPROVED** [review]
Verified with tests/boards/intel_adsp, tests/subsys/logging and tests/kernel and looks good to me.
